### DISCUSSION TITLE
`realm` new architecture adoption

### DISCUSSION THREAD

**Discussion by kraenhansen** [reactwg/react-native-new-architecture#95] [Libraries]
## TL;DR

- We have shipped `realm@11` (18th of October 2022) with support for JSC and Hermes through JSI. See [our blogpost for details](https://www.mongodb.com/developer/products/realm/realm-javascript-v11-react-native/).
- We're currently in progress with a complete rewrite of our SDK (to be released as v12 - when it's ready) and expect to adopt TurboModules in the process and some codegen in the process.

---

I hope this discussion will help shed some light on the ongoing efforts of the [Realm JS](https://github.com/realm/realm-js) team into supporting the new RN architecture.

## Historical context
- Our library has historically been very C++ heavy, since we're essentially publishing a binding between a database core library (written in C++) and every JS engine we support (JSC, V8 and now Hermes through JSI).
- We have had an abstraction over JS engines (a lot of C++ templating!) to solve the same problem that JSI eventually tackled.
- Our efforts to support running an app using Realm JS in the legacy Chrome debugger was slow and painful:
  - Our C++ had to function without any reliance on the JS which bootstrapped it when it was not in debugging mode.
  - We had to implement calls to the database through [a blocking HTTP call to an server running on device](https://github.com/realm/realm-js/blob/v10/lib/browser/rpc.js#L317).
   - Callback from C++ to JS was non-trivial to implement and the performance was poor as it involved polling from the client with an exponential backoff.

## Our JSI adoption & Hermes support

Our endeavors into the new architecture started as we implemented a JSI "backend" for our JS engine abstraction. We aimed to use this to replace our existing JSC "backend", as a JSC runtime was already available for JSI. This shipped the first time as `v10.20.0-alpha.0`, initially supporting iOS only, on the 1st of September 2021 and support for Android was introduced with `v10.20.0-alpha.1` on the 22nd of September 2021. We've fixed some bugs since, but at large our support for JSI hasn't been changing that much since then.

Most notably, we've seen [performance regressions](https://github.com/realm/realm-js/issues/3940#issuecomment-1238168460):
- Some is improved by [this hermes change](https://github.com/facebook/hermes/commit/2a32afb225349c75f3aeb4ffca10f3c4daab6fad).
- Some require [this feature request when end-users are running apps with Hermes disabled](https://github.com/react-native-community/discussions-and-proposals/issues/505).

We decided to drop support for the legacy chrome debugging mode, for a couple of reasons:
- As mentioned above, it made apps run [painfully slow](https://github.com/realm/realm-js/issues/491).
- Our implementation wasn't using our JS engine abstraction, but instead tightly coupled to JSC, which meant it would have to undergo a major refactoring.
- The Flipper plugin seemed to be catching adoption.
- When running JSC (not debuggable through Flipper) macOS users could work around [using Safari DevTools](https://blog.nparashuram.com/2019/10/debugging-react-native-ios-apps-with.html).

This has the unfortunate side-effect that users that haven't yet adopted Hermes and are not on a Mac, have no known way of debugging their apps if they're using Realm JS v11 (or later). Unfortunately this also includes dropping support for tools like the very popular [react-native-debugger](https://github.com/jhen0409/react-native-debugger). There's no doubt [we're betting on Flipper becoming the de-facto debugger for React Native developers](https://www.mongodb.com/developer/products/realm/realm-javascript-v11-react-native/).

## Our future

We're currently in progress with a complete rewrite of our SDK (to be released as v12 - when it's ready) and expect to adopt TurboModules in the process and some codegen in the process. We expect very few breaking changes to our external API, but a major overhaul of our internals. Most notably, we're now exposing a TypeScript binding for our internal database core library, which we expect will result in fewer bugs, less code to maintain and an increase in developer productivity on our team.

Since we're already generating a binding exposing our native C++ code via TypeScript, we expect very little leverage from the RN codegen tooling. We'll most likely expose a single method calling into C++ and returning an object with a lot of free functions that we bind into classes from JS code generated by our own binding generator.

I realize now this post got a bit longer than I expected: I hope this will inspire other maintainers and provide a bit of context for users of RN and our library. Please feel free to ask any questions for us regarding our adoption of the new architecture, here or more informally on the #realm channel on the "React Native Contributors" discord.

**Comment by kraenhansen** [comment]
Last week we released the first alpha release of our re-written SDK: https://github.com/realm/realm-js/releases/tag/v12.0.0-alpha.0
